runtime/error.c

```diff
@@ -148,15 +161,71 @@ void _bal_error_backtrace_print(ErrorPtr ep) {
     fflush(stderr);
 }
 
+static void processPCs(GC PC* pcs, uint32_t nPCs, uint32_t start, int64_t lineNumber, BacktraceError *error) {
+    // Handle the starting pc seperately
+    BacktraceStartLine backtraceStartLine = { { error->code, error->message }, lineNumber };
+    processInitialPC(pcs[start], lineNumber, &backtraceStartLine);
+    backtraceStartLine.lineNumber = 0;
+    if (backtraceStartLine.error.code == NO_DEBUG_INFO) {
+        for (uint32_t i = start + 1; i < nPCs; i++) {
+            // We do not need to pass error_callback because, already we have an error.
+            backtrace_syminfo(state, pcs[i], printMissingSymbols, NULL, &backtraceStartLine);
+        }
+    }
+    else {
+        for (uint32_t i = start + 1; i < nPCs; i++) {
+            backtrace_pcinfo(state, pcs[i], printBacktraceLine, storeBacktraceError, &backtraceStartLine);
+        }
+    }
+    error->code = backtraceStartLine.error.code;
+    error->message = backtraceStartLine.error.message;
+}
+
+static void processInitialPC(PC pc, int64_t lineNumber, BacktraceStartLine *backtraceStartLine) {
+    backtrace_pcinfo(state, pc, printBacktraceLine, storeBacktraceError, backtraceStartLine);
+    if (backtraceStartLine->error.code == NO_DEBUG_INFO) {
+        // We do not need to pass error_callback because, already we have an error
+        backtrace_syminfo(state, pc, printMissingSymbols, NULL, backtraceStartLine);
+    }
+}
+
 // Implementation of backtrace_full_callback
 static int printBacktraceLine(void *data, PC pc, const char *filename, int lineno, const char *function) {
```

Comment:
printBacktraceLine and printLine are not well named: printLine prints a line of a backtrace.

Maybe add `Callback` or `CB` at the end of the names of every function that is used as a callback for libbacktrace?

Then printBacktraceLine would become printBacktraceLineCB and printLine would become printBacktraceLine.

Reply:
Changed